assets/src/edit-story/app/history/historyProvider.js

```diff
@@ -36,18 +38,27 @@ function HistoryProvider({ children, size }) {
     undo,
     redo,
   } = useHistoryReducer(size);
+  const canUndo = offset < historyLength - 1;
+  const [hasChangedSinceLastSave, setHasChangedSinceLastSave] = useState(false);
+
+  usePreventWindowUnload(hasChangedSinceLastSave);
+
+  useEffect(() => {
+    setHasChangedSinceLastSave(canUndo);
+  }, [setHasChangedSinceLastSave, canUndo]);
```

Comment:
So, this will only run when these values change. But `canUndo` will stay the same after the first change after a save, no? So, making a change after saving will not trigger here? I think a better test would be to make sure it runs again for every change to the history (though saving will update the story properties which will in turn cause a change in history?).

I guess my question is: Does this actually work?

Reply:
Just fixed, we should store the records as references. Thanks for point this out.